src/cascadia/TerminalCore/TerminalSelection.cpp

```diff
@@ -341,6 +342,157 @@ void Terminal::SwitchSelectionEndpoint()
+    const auto viewportHeight = _GetMutableViewport().Height();
+
+    // extracts the next/previous hyperlink from the list of hyperlink ranges provided
+    auto extractResultFromList = [&, onUrl = _isTargetingUrl, selectionStart = _selection->start](const std::vector<interval_tree::Interval<til::point, size_t>>& list, bool forward, bool useViewportCoords) {
+        // checks if we're already on the hyperlink and if we're trying to set it to the same one
+        auto isAlreadySelected = [onUrl, selectionStartForValidation = useViewportCoords ? _ConvertToViewportCell(selectionStart) : selectionStart](const interval_tree::Interval<til::point, size_t> range) {
+            return onUrl && selectionStartForValidation == range.start;
+        };
+
+        std::optional<std::pair<til::point, til::point>> resultFromList;
+        if (list.size() == 1)
+        {
+            if (!isAlreadySelected(list.front()))
+            {
+                resultFromList = std::make_pair(list.front().start, list.front().stop);
+            }
+        }
+        else if (list.size() > 1)
+        {
+            // multiple results, find the first one in the direction we're going
+            if (forward)
+            {
+                for (const auto& range : list)
+                {
+                    if (isAlreadySelected(range))
+                    {
+                        continue;
+                    }
+                    else if (!resultFromList || range.start < resultFromList->first)
+                    {
+                        resultFromList = std::make_pair(range.start, range.stop);
+                    }
+                }
+            }
+            else
+            {
+                for (const auto& range : list)
+                {
+                    if (isAlreadySelected(range))
+                    {
+                        continue;
+                    }
+                    else if (!resultFromList || range.start > resultFromList->first)
+                    {
+                        resultFromList = std::make_pair(range.start, range.stop);
+                    }
+                }
+            }
+        }
+
+        // useViewportCoords --> pattern tree stores everything as viewport coords,
+        // so we need to convert them on the way out
+        if (useViewportCoords && resultFromList)
+        {
+            resultFromList->first = _ConvertToBufferCell(resultFromList->first);
+            resultFromList->second = _ConvertToBufferCell(resultFromList->second);
+        }
+        return resultFromList;
+    };
+
+    // 1. Look for the hyperlink
+    til::point searchStart = searchForward ? _selection->start : til::point{ bufferSize.Left(), _VisibleStartIndex() };
+    til::point searchEnd = searchForward ? til::point{ bufferSize.RightInclusive(), _VisibleEndIndex() } : _selection->start;
+
+    // 1.A) Try searching the current viewport (no scrolling required)
+    auto patterns = _patternIntervalTree;

```

Comment:
Yeah. I think that's necessary though? Here's the flow...
1. check current viewport (`patterns = _patternIntervalTree`). If it's in there, extract the result and we're done
2. check next viewport (`patterns = <get patterns from new search space>`)...
   - if one is found, we have to scroll there. Upon scrolling, we eventually refresh `_patternIntervalTree`
   - if none are found in the whole buffer, don't modify `_patternIntervalTree` in any way.

I guess the only way to avoid the copy would be to have a bool to track the first run, but that's not really worth it, right? Or am I missing something?

Reply:
But you only use `patterns` to call `findContained` right? Does the `pattern` copy need to live past the scrolling? It doesn't seem to me like it does. 🤔